Add some more regression tests for cookies

sysrqb opened the pull request:
#### 38118fcbd4f510983f40219a74caf5257139aaec
<pre>
Add some more regression tests for cookies
<a href="https://bugs.webkit.org/show_bug.cgi?id=297084">https://bugs.webkit.org/show_bug.cgi?id=297084</a>
<a href="https://rdar.apple.com/problem/157788048">rdar://problem/157788048</a>

Reviewed by Charlie Wolfe.

Add some more tests that cover cookie behavior. The tests are focused on
partitioned cookies, but it covers some other cookie behavior as a side-effect
because I want to make sure that unpartitioned cookies interact with
partitioned in the expected way.

There are some failing tests for two reasons. The first reason is related to
how the implementation handles first-party partitioned cookies. This may
require changes.

The second reason is because in one of the tests I am testing how cookies
behave within the first-party and third-party contexts, and the test doesn&apos;t
correctly decide pass/fail in some cases. I will update the test so the test
expectations show that the tests pass as expected in a follow up PR.

The single API test is a new regression for the cookieTransformCallback
behavior. I needed to allow insecure connections in TestWebKitAPI for this.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/clear-site-data/set-and-clear-cookies.https.html:
* LayoutTests/http/tests/cookies/accept-partitioned-first-and-third-party-cookies.https.html:
* LayoutTests/http/tests/cookies/max-partitioned-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/max-partitioned-cookies.https.html: Added.
* LayoutTests/http/tests/cookies/multiple-cookies-expected.txt:
* LayoutTests/http/tests/cookies/multiple-cookies-iframes.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/multiple-cookies-iframes.https.html: Added.
* LayoutTests/http/tests/cookies/multiple-cookies-with-partitioned.https.html:
* LayoutTests/http/tests/cookies/multiple-cookies.html:
* LayoutTests/http/tests/cookies/multiple-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/multiple-cookies.https.html: Added.
* LayoutTests/http/tests/cookies/only-accept-first-party-cookies-expected.txt:
* LayoutTests/http/tests/cookies/only-accept-first-party-cookies.html:
* LayoutTests/http/tests/cookies/resources/echo-cookies.py:
* LayoutTests/http/tests/cookies/resources/set-cookie-and-redirect-back.py:
* LayoutTests/http/tests/cookies/resources/set-cookie-and-serve.py:
* LayoutTests/http/tests/resourceLoadStatistics/only-accept-first-party-cookies-with-third-party-cookie-blocking-expected.txt: Copied from LayoutTests/http/tests/cookies/only-accept-first-party-cookies-expected.txt.
* LayoutTests/http/tests/resourceLoadStatistics/only-accept-first-party-cookies-with-third-party-cookie-blocking.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/only-partitioned-cookies-after-redirect-fetch.https-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/only-partitioned-cookies-after-redirect-fetch.https.html: Added.
* LayoutTests/http/tests/resources/redirect.py:
(set_cookie):
(add_cors):
* LayoutTests/http/tests/storageAccess/deny-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/storageAccess/deny-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https.html: Copied from LayoutTests/http/tests/storageAccess/grant-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https.html.

Similar to the other &quot;deny&quot; test, this won&apos;t work until we add support for rejecting the prompt.

* LayoutTests/http/tests/storageAccess/grant-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Tools/TestWebKitAPI/Info.plist:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(-[SessionDelegate URLSession:task:didReceiveChallenge:completionHandler:]):
(TEST(Navigation, CookieTransformOnRedirect)):

Canonical link: <a href="https://commits.webkit.org/298786@main">https://commits.webkit.org/298786@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e37f9881974b4e210adbb5d14ae573ae8c6849b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116633 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36297 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26859 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122707 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67205 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e756edaf-b06c-4c24-b966-378f3ffc4342) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118522 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44886 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88579 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b24a8216-7ec6-48e5-9585-4ce4bbb5bbe9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119582 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29505 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104627 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69045 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28569 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22734 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66374 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98885 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22892 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125843 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43532 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32703 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97251 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43896 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100829 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97044 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42367 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20291 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39494 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18624 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43418 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42884 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46224 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44590 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Tools/TestWebKitAPI/Info.plist:
```diff
@@ -30,5 +30,10 @@
+	<dict>
+		<key>NSAllowsArbitraryLoads</key>
+		<true/>
+	</dict>
```
Is this needed for the API test?

sysrqb replied:
Yes, I added a comment in the commit message, too.

charliewolfe commented on Source/WebCore/platform/network/NetworkStorageSession.h:
```diff
@@ -244,6 +244,7 @@ class NetworkStorageSession
     ThirdPartyCookieBlockingDecision thirdPartyCookieBlockingDecisionForRequest(const URL& firstPartyForCookies, const URL& resource, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, ShouldRelaxThirdPartyCookieBlocking) const;
     WEBCORE_EXPORT bool shouldBlockCookies(const ResourceRequest&, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, ShouldRelaxThirdPartyCookieBlocking) const;
     WEBCORE_EXPORT bool shouldBlockCookies(const URL& firstPartyForCookies, const URL& resource, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, ShouldRelaxThirdPartyCookieBlocking) const;
+    WEBCORE_EXPORT bool shouldBlockCookies(const URL& firstPartyForCookies, const URL& resource, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, ShouldRelaxThirdPartyCookieBlocking) const;
```
I don't think you meant to add this

sysrqb replied:
ugh. no

charliewolfe reviewed (commented):
I don't follow what the purpose of these tests is. Are they all related to CHIPS, or are they just various cookie tests? Why are some of the tests failing? Should we fix those failures?

It would be helpful if there was a bit more information in the commit message.

charliewolfe reviewed (approved):
One of the tests needs to be skipped on glib

Commit:
Add some more regression tests for cookies
https://bugs.webkit.org/show_bug.cgi?id=297084
rdar://problem/157788048

Reviewed by Charlie Wolfe.

Add some more tests that cover cookie behavior. The tests are focused on
partitioned cookies, but it covers some other cookie behavior as a side-effect
because I want to make sure that unpartitioned cookies interact with
partitioned in the expected way.

There are some failing tests for two reasons. The first reason is related to
how the implementation handles first-party partitioned cookies. This may
require changes.

The second reason is because in one of the tests I am testing how cookies
behave within the first-party and third-party contexts, and the test doesn't
correctly decide pass/fail in some cases. I will update the test so the test
expectations show that the tests pass as expected in a follow up PR.

The single API test is a new regression for the cookieTransformCallback
behavior. I needed to allow insecure connections in TestWebKitAPI for this.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/clear-site-data/set-and-clear-cookies.https.html:
* LayoutTests/http/tests/cookies/accept-partitioned-first-and-third-party-cookies.https.html:
* LayoutTests/http/tests/cookies/max-partitioned-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/max-partitioned-cookies.https.html: Added.
* LayoutTests/http/tests/cookies/multiple-cookies-expected.txt:
* LayoutTests/http/tests/cookies/multiple-cookies-iframes.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/multiple-cookies-iframes.https.html: Added.
* LayoutTests/http/tests/cookies/multiple-cookies-with-partitioned.https.html:
* LayoutTests/http/tests/cookies/multiple-cookies.html:
* LayoutTests/http/tests/cookies/multiple-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/multiple-cookies.https.html: Added.
* LayoutTests/http/tests/cookies/only-accept-first-party-cookies-expected.txt:
* LayoutTests/http/tests/cookies/only-accept-first-party-cookies.html:
* LayoutTests/http/tests/cookies/resources/echo-cookies.py:
* LayoutTests/http/tests/cookies/resources/set-cookie-and-redirect-back.py:
* LayoutTests/http/tests/cookies/resources/set-cookie-and-serve.py:
* LayoutTests/http/tests/resourceLoadStatistics/only-accept-first-party-cookies-with-third-party-cookie-blocking-expected.txt: Copied from LayoutTests/http/tests/cookies/only-accept-first-party-cookies-expected.txt.
* LayoutTests/http/tests/resourceLoadStatistics/only-accept-first-party-cookies-with-third-party-cookie-blocking.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/only-partitioned-cookies-after-redirect-fetch.https-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/only-partitioned-cookies-after-redirect-fetch.https.html: Added.
* LayoutTests/http/tests/resources/redirect.py:
(set_cookie):
(add_cors):
* LayoutTests/http/tests/storageAccess/deny-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/storageAccess/deny-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https.html: Copied from LayoutTests/http/tests/storageAccess/grant-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https.html.

Similar to the other "deny" test, this won't work until we add support for rejecting the prompt.

* LayoutTests/http/tests/storageAccess/grant-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Tools/TestWebKitAPI/Info.plist:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(-[SessionDelegate URLSession:task:didReceiveChallenge:completionHandler:]):
(TEST(Navigation, CookieTransformOnRedirect)):

Canonical link: https://commits.webkit.org/298786@main
